incubator/hnc/internal/reconcilers/object_test.go

```diff
@@ -15,6 +15,201 @@ import (
+				}
+				makeObjectWithAnnotation(ctx, "Role", fooName, "testrole", a)
+				for _, ns := range tc.want {
+					ns = replaceStrings(ns, indToNm)
+					Eventually(hasObject(ctx, "Role", ns, "testrole")).Should(BeTrue(), "When propagating testrole to %s", ns)
+				}
+				for _, ns := range tc.notWant {
+					ns = replaceStrings(ns, indToNm)
+					Consistently(hasObject(ctx, "Role", ns, "testrole")).Should(BeFalse(), "When propagating testrole to %s", ns)
+				}
+			})
+		}
+	})
+
+	Context("Update exception annotations ", func() {
+		const (
+			p  = "parent"
+			c1 = "child1"
+			c2 = "child2"
+			c3 = "child3"
+		)
+		tests := []struct {
+			name         string
+			selector     string
+			treeSelector string
+			want         []string
+			notWant      []string
+		}{
+			{name: "update select",
+				selector:     "!" + c1 + api.LabelTreeDepthSuffix,
+				treeSelector: "",
+				want:         []string{c2, c3},
+				notWant:      []string{c1},
+			},
+			{name: "update treeSelect",
+				selector:     "",
+				treeSelector: "!" + c1,
+				want:         []string{c2, c3},
+				notWant:      []string{c1},
+			},
+		}
+
+		for _, tc := range tests {
+			// Making a local copy of tc is necessary to ensure the correct value is passed to the closure,
+			// for more details look at https://onsi.github.io/ginkgo/ and search for 'closure'
+			tc := tc
+			It("Should handle exceptions: "+tc.name, func() {
+				// Set up namespaces
+				fooName := createNS(ctx, "foo")
+				barName := createNS(ctx, "bar")
+				bazName := createNS(ctx, "baz")
+				quxName := createNS(ctx, "qux")
+
+				setParent(ctx, barName, fooName)
+				setParent(ctx, bazName, fooName)
+				setParent(ctx, quxName, fooName)
+
+				// Convert namespace index to real names
+				indToNm := map[string]string{
+					p:  fooName,
+					c1: barName,
+					c2: bazName,
+					c3: quxName,
+				}
+				tc.selector = replaceStrings(tc.selector, indToNm)
+				tc.treeSelector = replaceStrings(tc.treeSelector, indToNm)
+
+				// Create a Role and verify it's propagated
+				makeObject(ctx, "Role", fooName, "testrole")
+				for _, ns := range []string{barName, bazName, quxName} {
+					Eventually(hasObject(ctx, "Role", ns, "testrole")).Should(BeTrue(), "When propagating testrole to %s", ns)
+				}
+
+				// update the role with the selector and treeSelector annotation
+				a := make(map[string]string)
+				if tc.selector != "" {
+					a["propagate.hnc.x-k8s.io/select"] = tc.selector
+				}
+				if tc.treeSelector != "" {
+					a["propagate.hnc.x-k8s.io/treeSelect"] = tc.treeSelector
+				}
+				updateObjectWithAnnotation(ctx, "Role", fooName, "testrole", a)
+				// make sure the changes are propagated
+				for _, ns := range tc.notWant {
+					ns = replaceStrings(ns, indToNm)
+					Eventually(hasObject(ctx, "Role", ns, "testrole")).Should(BeFalse(), "When propagating testrole to %s", ns)
+				}
+				// then check that the objects are still reserved in these namespaces
```

Comment:
"reserved"?

Also it might be nice to try _removing_ the annotation and verify that the objects are re-propagated everywhere

Reply:
done